{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Segmentation free"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import scanpy as sc\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "### directory & filepaths\n",
    "data_dir = '/home/shamini/data1/data_orig/data/spatial/xenium/10xGenomics/cell_seg_brain_cancer/'\n",
    "out = '/home/shamini/data/projects/spatial_workshop/out/'\n",
    "os.makedirs(out+'module6/script06_figures/', exist_ok=True)\n",
    "os.makedirs(out+'module6/script06_objects/segmentation/', exist_ok=True)\n",
    "os.makedirs(out+'module6/script06_objects/ncv/', exist_ok=True)\n",
    "os.makedirs(out+'module6/script06_objects/segmentation_with_cell_centers/', exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_crop = sc.read_h5ad(out+\"module6/adata_crop.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_crop\n",
    "\n",
    "### downsample transcripts_df\n",
    "\n",
    "transcripts_df = pd.read_csv(out+'module6/transcripts_subset_all_genes.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_centers = adata_crop.uns['spatial']\n",
    "cell_centers\n",
    "\n",
    "transcripts_df_downsampled = transcripts_df.sample(frac=0.001)\n",
    "transcripts_df_downsampled.to_csv(out+'module6/script06_objects/transcripts_df_downsampled.csv', index=False)\n",
    "cell_centers_downsampled = cell_centers[cell_centers.index.isin(transcripts_df_downsampled['cell_id'])]\n",
    "cell_centers_downsampled.to_csv(out+'module6/script06_objects/cell_centers_downsampled.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[04:13:27] Info: Run Rb1d7ec66e\n",
      "[04:13:27] Info: (2024-06-30) Run Baysor v0.6.2\n",
      "[04:13:27] Info: Loading data...\n",
      "[04:13:29] Info: Loaded 1503 transcripts\n",
      "[04:13:31] Info: Estimating noise level\n",
      "[04:13:33] Info: Done\n",
      "[04:13:47] Info: Clustering molecules...\n",
      "[04:13:48] Warning: ICA did not converge, fall back to random initialization\n",
      "└ Baysor.Processing /home/viktor_petukhov/.julia/dev/Baysor/src/processing/bmm_algorithm/molecule_clustering.jl:147\n",
      "[04:13:49] Info: Algorithm stopped after 21 iterations. Error: 5.04e-5. Converged: true.\n",
      "[04:13:49] Info: Done\n",
      "[04:13:49] Info: Initializing algorithm. Scale: 9.0, scale std: 2.25, initial #components: 120, #molecules: 1503.\n",
      "[04:13:49] Info: Using the following additional information about molecules: [:confidence, :cluster]\n",
      "[04:13:49] Info: Using 3D coordinates\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Progress:  85%|██████████████████████████████████▉      |  ETA: 0:00:00\u001b[K\n",
      "  Iteration:        23\u001b[K\n",
      "  Noise level, %:   32.27\u001b[K\n",
      "\u001b[Aum. components:  0\u001b[K\n",
      "\n",
      "\n",
      "Progress: 100%|█████████████████████████████████████████| Time: 0:00:00\u001b[K\n",
      "  Iteration:        27\u001b[K\n",
      "  Noise level, %:   32.07\u001b[K\n",
      "  Num. components:  0\u001b[K\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[04:13:50] Info: Processing complete.\n",
      "[04:13:50] Info: Estimating local colors\n",
      "[04:14:04] Warning: n=1503, which is > length(high_conf_ids) (1474)\n",
      "└ Baysor.Processing /home/viktor_petukhov/.julia/dev/Baysor/src/processing/data_processing/initialization.jl:180\n",
      "[04:14:17] Info: Estimating boundary polygons\n",
      "[04:14:17] Warning: To many values of z. Using 2D polygons\n",
      "└ Baysor.Processing /home/viktor_petukhov/.julia/dev/Baysor/src/processing/data_processing/boundary_estimation.jl:143\n",
      "[04:14:17] Info: Saving results to /home/shamini/data/projects/spatial_workshop/out/module6/script06_objects/segmentation/segmentation.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Progress: 100%|█████████████████████████████████████████| Time: 0:00:00\u001b[K\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[04:14:24] Info: All done!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['/home/shamini/baysor/bin/baysor/bin/baysor', 'run', '-c', '/home/shamini/data/projects/spatial_workshop/scripts/module6/parameters_baysor.toml', '-s', '9', '--save-polygons', 'geojson', '-o', '/home/shamini/data/projects/spatial_workshop/out/module6/script06_objects/segmentation/', '/home/shamini/data/projects/spatial_workshop/out/module6/script06_objects/transcripts_df_downsampled.csv'], returncode=0)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# Define the Baysor command with the --save-polygons flag\n",
    "\n",
    "# get filtered transcript files\n",
    "### Read in the transcripts_df.csv file from module 2\n",
    "\n",
    "command = [\n",
    "        '/home/shamini/baysor/bin/baysor/bin/baysor', 'run',\n",
    "        '-c', '/home/shamini/data/projects/spatial_workshop/scripts/module6/parameters_baysor.toml',\n",
    "        '-s', '9',\n",
    "        '--save-polygons', 'geojson',\n",
    "        '-o', os.path.join(out, 'module6/script06_objects/segmentation/'),\n",
    "        os.path.join(out,'module6/script06_objects/transcripts_df_downsampled.csv')\n",
    "        ]\n",
    "subprocess.run(command, check=True)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[04:14:33] Info: Run R4dfe4183a\n",
      "[04:14:33] Info: # CLI params: `-c /home/shamini/data/projects/spatial_workshop/scripts/module6/parameters_baysor.toml -o /home/shamini/data/projects/spatial_workshop/out/module6/script06_objects/ncv/ /home/shamini/data/projects/spatial_workshop/out/module6/script06_objects/transcripts_df_downsampled.csv`\n",
      "[04:14:33] Info: (2024-06-30) Run Baysor v0.6.2\n",
      "[04:14:33] Info: Loading data...\n",
      "[04:14:35] Info: Loaded 1503 transcripts\n",
      "[04:14:36] Info: Estimating neighborhoods...\n",
      "[04:14:36] Info: Estimating molecule confidences...\n",
      "[04:14:37] Info: Estimating gene colors...\n",
      "[04:14:57] Warning: n=1503, which is > length(high_conf_ids) (1474)\n",
      "└ Baysor.Processing /home/viktor_petukhov/.julia/dev/Baysor/src/processing/data_processing/initialization.jl:180\n",
      "[04:15:09] Info: Saving results...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Progress: 100%|█████████████████████████████████████████| Time: 0:00:00\u001b[K\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[04:15:14] Info: Done!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['/home/shamini/baysor/bin/baysor/bin/baysor', 'segfree', '-c', '/home/shamini/data/projects/spatial_workshop/scripts/module6/parameters_baysor.toml', '-o', '/home/shamini/data/projects/spatial_workshop/out/module6/script06_objects/ncv/', '/home/shamini/data/projects/spatial_workshop/out/module6/script06_objects/transcripts_df_downsampled.csv'], returncode=0)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "command = [\n",
    "        '/home/shamini/baysor/bin/baysor/bin/baysor', 'segfree',\n",
    "        '-c', '/home/shamini/data/projects/spatial_workshop/scripts/module6/parameters_baysor.toml',\n",
    "        '-o', os.path.join(out, 'module6/script06_objects/ncv/'),\n",
    "        os.path.join(out,'module6/script06_objects/transcripts_df_downsampled.csv')\n",
    "        ]\n",
    "subprocess.run(command, check=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### load baysor output\n",
    "baysor_seg = pd.read_csv(out+'module6/script06_objects/segmentation/segmentation.csv')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>cell_id</th>\n",
       "      <th>overlaps_nucleus</th>\n",
       "      <th>gene</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>z</th>\n",
       "      <th>qv</th>\n",
       "      <th>fov_name</th>\n",
       "      <th>nucleus_distance</th>\n",
       "      <th>codeword_index</th>\n",
       "      <th>molecule_id</th>\n",
       "      <th>confidence</th>\n",
       "      <th>cluster</th>\n",
       "      <th>cell</th>\n",
       "      <th>assignment_confidence</th>\n",
       "      <th>is_noise</th>\n",
       "      <th>ncv_color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>161376389</td>\n",
       "      <td>282462819813394</td>\n",
       "      <td>jpaamlpc-1</td>\n",
       "      <td>0</td>\n",
       "      <td>SPARCL1</td>\n",
       "      <td>365.15625</td>\n",
       "      <td>878.8574</td>\n",
       "      <td>19.536585</td>\n",
       "      <td>10.535770</td>\n",
       "      <td>Z17</td>\n",
       "      <td>2.188293</td>\n",
       "      <td>378</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.66667</td>\n",
       "      <td>True</td>\n",
       "      <td>#0077A6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>170285965</td>\n",
       "      <td>281513632243259</td>\n",
       "      <td>joffddop-1</td>\n",
       "      <td>0</td>\n",
       "      <td>TUBA1B</td>\n",
       "      <td>4032.10550</td>\n",
       "      <td>3725.6094</td>\n",
       "      <td>23.619045</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>AA19</td>\n",
       "      <td>0.581290</td>\n",
       "      <td>185</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>CRb1d7ec66e-162</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>False</td>\n",
       "      <td>#408594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>164489968</td>\n",
       "      <td>282467114474721</td>\n",
       "      <td>amnedlkf-1</td>\n",
       "      <td>0</td>\n",
       "      <td>STAT3</td>\n",
       "      <td>2163.27730</td>\n",
       "      <td>1525.0566</td>\n",
       "      <td>23.228128</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>Z18</td>\n",
       "      <td>1.907755</td>\n",
       "      <td>179</td>\n",
       "      <td>3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>CRb1d7ec66e-192</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>False</td>\n",
       "      <td>#00A100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>161356570</td>\n",
       "      <td>282462819672451</td>\n",
       "      <td>jhacgahf-1</td>\n",
       "      <td>1</td>\n",
       "      <td>SPARCL1</td>\n",
       "      <td>326.72656</td>\n",
       "      <td>1168.2207</td>\n",
       "      <td>24.124050</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>Z17</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>378</td>\n",
       "      <td>4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>CRb1d7ec66e-57</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>False</td>\n",
       "      <td>#0078B5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>167088718</td>\n",
       "      <td>282467114626591</td>\n",
       "      <td>jplafcde-1</td>\n",
       "      <td>1</td>\n",
       "      <td>CD74</td>\n",
       "      <td>3140.55080</td>\n",
       "      <td>1123.8848</td>\n",
       "      <td>21.005756</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>Z18</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>364</td>\n",
       "      <td>5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>CRb1d7ec66e-260</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>False</td>\n",
       "      <td>#00A455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1498</th>\n",
       "      <td>164455770</td>\n",
       "      <td>282467114235027</td>\n",
       "      <td>jpidnilh-1</td>\n",
       "      <td>1</td>\n",
       "      <td>ACTB</td>\n",
       "      <td>1719.72270</td>\n",
       "      <td>1464.1309</td>\n",
       "      <td>23.411846</td>\n",
       "      <td>39.587856</td>\n",
       "      <td>Z18</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>373</td>\n",
       "      <td>1499</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>True</td>\n",
       "      <td>#009A81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1499</th>\n",
       "      <td>169982721</td>\n",
       "      <td>282471409232755</td>\n",
       "      <td>jlmgchdh-1</td>\n",
       "      <td>1</td>\n",
       "      <td>SLC1A3</td>\n",
       "      <td>4192.46880</td>\n",
       "      <td>1692.1914</td>\n",
       "      <td>26.176071</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>Z19</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>216</td>\n",
       "      <td>1500</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>CRb1d7ec66e-38</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>False</td>\n",
       "      <td>#6C8D5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1500</th>\n",
       "      <td>170239015</td>\n",
       "      <td>281513631691370</td>\n",
       "      <td>jliemifl-1</td>\n",
       "      <td>1</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>4538.48440</td>\n",
       "      <td>3934.4688</td>\n",
       "      <td>23.799513</td>\n",
       "      <td>23.529787</td>\n",
       "      <td>AA19</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>94</td>\n",
       "      <td>1501</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>CRb1d7ec66e-217</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>False</td>\n",
       "      <td>#009DBE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1501</th>\n",
       "      <td>164538243</td>\n",
       "      <td>282467114832449</td>\n",
       "      <td>jphghcei-1</td>\n",
       "      <td>1</td>\n",
       "      <td>CDKN2C</td>\n",
       "      <td>1944.68750</td>\n",
       "      <td>1260.3848</td>\n",
       "      <td>21.475980</td>\n",
       "      <td>18.697315</td>\n",
       "      <td>Z18</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1502</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.66667</td>\n",
       "      <td>True</td>\n",
       "      <td>#007A4C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1502</th>\n",
       "      <td>167325200</td>\n",
       "      <td>281509336884920</td>\n",
       "      <td>jkejfgfh-1</td>\n",
       "      <td>0</td>\n",
       "      <td>ARPC5</td>\n",
       "      <td>2799.04690</td>\n",
       "      <td>2524.8320</td>\n",
       "      <td>22.975636</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>AA18</td>\n",
       "      <td>0.885713</td>\n",
       "      <td>213</td>\n",
       "      <td>1503</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>CRb1d7ec66e-129</td>\n",
       "      <td>0.66667</td>\n",
       "      <td>False</td>\n",
       "      <td>#007B57</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1503 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0    transcript_id     cell_id  overlaps_nucleus     gene  \\\n",
       "0      161376389  282462819813394  jpaamlpc-1                 0  SPARCL1   \n",
       "1      170285965  281513632243259  joffddop-1                 0   TUBA1B   \n",
       "2      164489968  282467114474721  amnedlkf-1                 0    STAT3   \n",
       "3      161356570  282462819672451  jhacgahf-1                 1  SPARCL1   \n",
       "4      167088718  282467114626591  jplafcde-1                 1     CD74   \n",
       "...          ...              ...         ...               ...      ...   \n",
       "1498   164455770  282467114235027  jpidnilh-1                 1     ACTB   \n",
       "1499   169982721  282471409232755  jlmgchdh-1                 1   SLC1A3   \n",
       "1500   170239015  281513631691370  jliemifl-1                 1     EGFR   \n",
       "1501   164538243  282467114832449  jphghcei-1                 1   CDKN2C   \n",
       "1502   167325200  281509336884920  jkejfgfh-1                 0    ARPC5   \n",
       "\n",
       "               x          y          z         qv fov_name  nucleus_distance  \\\n",
       "0      365.15625   878.8574  19.536585  10.535770      Z17          2.188293   \n",
       "1     4032.10550  3725.6094  23.619045  40.000000     AA19          0.581290   \n",
       "2     2163.27730  1525.0566  23.228128  40.000000      Z18          1.907755   \n",
       "3      326.72656  1168.2207  24.124050  40.000000      Z17          0.000000   \n",
       "4     3140.55080  1123.8848  21.005756  40.000000      Z18          0.000000   \n",
       "...          ...        ...        ...        ...      ...               ...   \n",
       "1498  1719.72270  1464.1309  23.411846  39.587856      Z18          0.000000   \n",
       "1499  4192.46880  1692.1914  26.176071  40.000000      Z19          0.000000   \n",
       "1500  4538.48440  3934.4688  23.799513  23.529787     AA19          0.000000   \n",
       "1501  1944.68750  1260.3848  21.475980  18.697315      Z18          0.000000   \n",
       "1502  2799.04690  2524.8320  22.975636  40.000000     AA18          0.885713   \n",
       "\n",
       "      codeword_index  molecule_id  confidence  cluster             cell  \\\n",
       "0                378            1         1.0        9              NaN   \n",
       "1                185            2         1.0        1  CRb1d7ec66e-162   \n",
       "2                179            3         1.0        1  CRb1d7ec66e-192   \n",
       "3                378            4         1.0        2   CRb1d7ec66e-57   \n",
       "4                364            5         1.0        1  CRb1d7ec66e-260   \n",
       "...              ...          ...         ...      ...              ...   \n",
       "1498             373         1499         1.0        1              NaN   \n",
       "1499             216         1500         1.0        1   CRb1d7ec66e-38   \n",
       "1500              94         1501         1.0        2  CRb1d7ec66e-217   \n",
       "1501               0         1502         1.0        9              NaN   \n",
       "1502             213         1503         1.0        1  CRb1d7ec66e-129   \n",
       "\n",
       "      assignment_confidence  is_noise ncv_color  \n",
       "0                   0.66667      True   #0077A6  \n",
       "1                   1.00000     False   #408594  \n",
       "2                   1.00000     False   #00A100  \n",
       "3                   1.00000     False   #0078B5  \n",
       "4                   1.00000     False   #00A455  \n",
       "...                     ...       ...       ...  \n",
       "1498                1.00000      True   #009A81  \n",
       "1499                1.00000     False   #6C8D5D  \n",
       "1500                1.00000     False   #009DBE  \n",
       "1501                0.66667      True   #007A4C  \n",
       "1502                0.66667     False   #007B57  \n",
       "\n",
       "[1503 rows x 19 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "baysor_seg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Baysor segmentation results in 1182 number of uique cells\n",
      " Original number of cells in the downsampled transcript file is 1529\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print(f\" Baysor segmentation results in {baysor_seg['cell_id'].nunique()} number of uique cells\")\n",
    "print(f\" Original number of cells in the downsampled transcript file is {transcripts_df_downsampled['cell_id'].nunique()}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>cell_id</th>\n",
       "      <th>overlaps_nucleus</th>\n",
       "      <th>gene</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>z</th>\n",
       "      <th>qv</th>\n",
       "      <th>fov_name</th>\n",
       "      <th>nucleus_distance</th>\n",
       "      <th>codeword_index</th>\n",
       "      <th>molecule_id</th>\n",
       "      <th>confidence</th>\n",
       "      <th>cluster</th>\n",
       "      <th>cell</th>\n",
       "      <th>assignment_confidence</th>\n",
       "      <th>is_noise</th>\n",
       "      <th>ncv_color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>170288686</td>\n",
       "      <td>281513632250980</td>\n",
       "      <td>jlgdpmln-1</td>\n",
       "      <td>0</td>\n",
       "      <td>STMN1</td>\n",
       "      <td>4678.37100</td>\n",
       "      <td>4585.75800</td>\n",
       "      <td>24.801740</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>AA19</td>\n",
       "      <td>0.291992</td>\n",
       "      <td>345</td>\n",
       "      <td>31</td>\n",
       "      <td>0.00027</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>#00A442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>169868892</td>\n",
       "      <td>282471409328081</td>\n",
       "      <td>jloaclbb-1</td>\n",
       "      <td>0</td>\n",
       "      <td>STMN1</td>\n",
       "      <td>4382.53500</td>\n",
       "      <td>638.55664</td>\n",
       "      <td>22.475628</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>Z19</td>\n",
       "      <td>1.055136</td>\n",
       "      <td>345</td>\n",
       "      <td>32</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>2</td>\n",
       "      <td>CRb1d7ec66e-200</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>#0DFFA6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>167318505</td>\n",
       "      <td>281509336839293</td>\n",
       "      <td>jkeaphmg-1</td>\n",
       "      <td>1</td>\n",
       "      <td>STMN1</td>\n",
       "      <td>2711.62500</td>\n",
       "      <td>2441.72660</td>\n",
       "      <td>21.385849</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>AA18</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>345</td>\n",
       "      <td>54</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>2</td>\n",
       "      <td>CRb1d7ec66e-282</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>#007B59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>169919658</td>\n",
       "      <td>282471409874279</td>\n",
       "      <td>jlclcakp-1</td>\n",
       "      <td>1</td>\n",
       "      <td>STMN1</td>\n",
       "      <td>4701.07400</td>\n",
       "      <td>1076.81050</td>\n",
       "      <td>24.665497</td>\n",
       "      <td>21.592703</td>\n",
       "      <td>Z19</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>345</td>\n",
       "      <td>62</td>\n",
       "      <td>0.95469</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>#1AFDA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>161570125</td>\n",
       "      <td>281505041666840</td>\n",
       "      <td>jkceplfn-1</td>\n",
       "      <td>1</td>\n",
       "      <td>ANXA1</td>\n",
       "      <td>711.52344</td>\n",
       "      <td>2459.73630</td>\n",
       "      <td>22.572563</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>AA17</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>336</td>\n",
       "      <td>95</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>9</td>\n",
       "      <td>CRb1d7ec66e-11</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>#005000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>167040538</td>\n",
       "      <td>282467114291552</td>\n",
       "      <td>jljmlhal-1</td>\n",
       "      <td>1</td>\n",
       "      <td>STMN1</td>\n",
       "      <td>3459.95700</td>\n",
       "      <td>625.56836</td>\n",
       "      <td>24.879513</td>\n",
       "      <td>27.992680</td>\n",
       "      <td>Z18</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>345</td>\n",
       "      <td>196</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>#00C078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>167185181</td>\n",
       "      <td>282467114390393</td>\n",
       "      <td>jpkbpgin-1</td>\n",
       "      <td>0</td>\n",
       "      <td>STMN1</td>\n",
       "      <td>2708.38670</td>\n",
       "      <td>1283.14060</td>\n",
       "      <td>23.102010</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>Z18</td>\n",
       "      <td>1.487329</td>\n",
       "      <td>345</td>\n",
       "      <td>203</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>2</td>\n",
       "      <td>CRb1d7ec66e-108</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>#00A129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>164520737</td>\n",
       "      <td>282467114697534</td>\n",
       "      <td>jphhgaon-1</td>\n",
       "      <td>1</td>\n",
       "      <td>STMN1</td>\n",
       "      <td>2296.15230</td>\n",
       "      <td>1435.29880</td>\n",
       "      <td>20.897295</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>Z18</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>345</td>\n",
       "      <td>218</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>2</td>\n",
       "      <td>CRb1d7ec66e-277</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>#00A403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>161450931</td>\n",
       "      <td>282462819382327</td>\n",
       "      <td>jgibdlce-1</td>\n",
       "      <td>1</td>\n",
       "      <td>ANXA1</td>\n",
       "      <td>293.65234</td>\n",
       "      <td>2092.77540</td>\n",
       "      <td>20.351965</td>\n",
       "      <td>23.087343</td>\n",
       "      <td>Z17</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>336</td>\n",
       "      <td>257</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>7</td>\n",
       "      <td>CRb1d7ec66e-176</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>#005B00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273</th>\n",
       "      <td>161405120</td>\n",
       "      <td>282462820009886</td>\n",
       "      <td>UNASSIGNED</td>\n",
       "      <td>0</td>\n",
       "      <td>STMN1</td>\n",
       "      <td>168.21094</td>\n",
       "      <td>1042.07810</td>\n",
       "      <td>19.839382</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>Z17</td>\n",
       "      <td>6.684786</td>\n",
       "      <td>345</td>\n",
       "      <td>274</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>2</td>\n",
       "      <td>CRb1d7ec66e-126</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>#0079B5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Unnamed: 0    transcript_id     cell_id  overlaps_nucleus   gene  \\\n",
       "30    170288686  281513632250980  jlgdpmln-1                 0  STMN1   \n",
       "31    169868892  282471409328081  jloaclbb-1                 0  STMN1   \n",
       "53    167318505  281509336839293  jkeaphmg-1                 1  STMN1   \n",
       "61    169919658  282471409874279  jlclcakp-1                 1  STMN1   \n",
       "94    161570125  281505041666840  jkceplfn-1                 1  ANXA1   \n",
       "195   167040538  282467114291552  jljmlhal-1                 1  STMN1   \n",
       "202   167185181  282467114390393  jpkbpgin-1                 0  STMN1   \n",
       "217   164520737  282467114697534  jphhgaon-1                 1  STMN1   \n",
       "256   161450931  282462819382327  jgibdlce-1                 1  ANXA1   \n",
       "273   161405120  282462820009886  UNASSIGNED                 0  STMN1   \n",
       "\n",
       "              x           y          z         qv fov_name  nucleus_distance  \\\n",
       "30   4678.37100  4585.75800  24.801740  40.000000     AA19          0.291992   \n",
       "31   4382.53500   638.55664  22.475628  40.000000      Z19          1.055136   \n",
       "53   2711.62500  2441.72660  21.385849  40.000000     AA18          0.000000   \n",
       "61   4701.07400  1076.81050  24.665497  21.592703      Z19          0.000000   \n",
       "94    711.52344  2459.73630  22.572563  40.000000     AA17          0.000000   \n",
       "195  3459.95700   625.56836  24.879513  27.992680      Z18          0.000000   \n",
       "202  2708.38670  1283.14060  23.102010  40.000000      Z18          1.487329   \n",
       "217  2296.15230  1435.29880  20.897295  40.000000      Z18          0.000000   \n",
       "256   293.65234  2092.77540  20.351965  23.087343      Z17          0.000000   \n",
       "273   168.21094  1042.07810  19.839382  40.000000      Z17          6.684786   \n",
       "\n",
       "     codeword_index  molecule_id  confidence  cluster             cell  \\\n",
       "30              345           31     0.00027        2              NaN   \n",
       "31              345           32     1.00000        2  CRb1d7ec66e-200   \n",
       "53              345           54     1.00000        2  CRb1d7ec66e-282   \n",
       "61              345           62     0.95469        2              NaN   \n",
       "94              336           95     1.00000        9   CRb1d7ec66e-11   \n",
       "195             345          196     1.00000        2              NaN   \n",
       "202             345          203     1.00000        2  CRb1d7ec66e-108   \n",
       "217             345          218     1.00000        2  CRb1d7ec66e-277   \n",
       "256             336          257     1.00000        7  CRb1d7ec66e-176   \n",
       "273             345          274     1.00000        2  CRb1d7ec66e-126   \n",
       "\n",
       "     assignment_confidence  is_noise ncv_color  \n",
       "30                     1.0      True   #00A442  \n",
       "31                     1.0     False   #0DFFA6  \n",
       "53                     1.0     False   #007B59  \n",
       "61                     1.0      True   #1AFDA2  \n",
       "94                     1.0     False   #005000  \n",
       "195                    1.0      True   #00C078  \n",
       "202                    1.0     False   #00A129  \n",
       "217                    1.0     False   #00A403  \n",
       "256                    1.0     False   #005B00  \n",
       "273                    1.0     False   #0079B5  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### subset the same genes as before , PTPCR, ANXA1, STMN1\n",
    "\n",
    "genes = ['PTPRC', 'ANXA1', 'STMN1']\n",
    "baysor_seg_genes = baysor_seg[baysor_seg['gene'].isin(genes)]\n",
    "baysor_seg_genes.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Number of cells expressing all 3 genes is 0\n",
      " Number of cells expressing STMN1 and ANXA1 is 0\n",
      " Number of cells expressing STMN1 and PTPRC is 0\n",
      " Number of cells expressing ANXA1 and PTPRC is 0\n",
      " Number of cells expressing STMN1 is 31\n",
      " Number of cells expressing ANXA1 is 9\n",
      " Number of cells expressing PTPRC is 0\n"
     ]
    }
   ],
   "source": [
    "\n",
    "cells_stmn1 = baysor_seg_genes['cell_id'][baysor_seg_genes['gene']=='STMN1'].unique()\n",
    "cells_anxa1 = baysor_seg_genes['cell_id'][baysor_seg_genes['gene']=='ANXA1'].unique()\n",
    "cells_ptprc = baysor_seg_genes['cell_id'][baysor_seg_genes['gene']=='PTPRC'].unique()\n",
    "\n",
    "### cells expressing all 3 genes\n",
    "cells_all = set(cells_stmn1) & set(cells_anxa1) & set(cells_ptprc)\n",
    "\n",
    "### cells expressing stmn1 & anxa1\n",
    "cells_stmn1_anxa1 = set(cells_stmn1) & set(cells_anxa1)\n",
    "\n",
    "### cells expressing stmn1 & ptprc\n",
    "cells_stmn1_ptprc = set(cells_stmn1) & set(cells_ptprc)\n",
    "\n",
    "### cells expressing anxa1 & ptprc\n",
    "cells_anxa1_ptprc = set(cells_anxa1) & set(cells_ptprc)\n",
    "\n",
    "print(f\" Number of cells expressing all 3 genes is {len(cells_all)}\")\n",
    "print(f\" Number of cells expressing STMN1 and ANXA1 is {len(cells_stmn1_anxa1)}\")\n",
    "print(f\" Number of cells expressing STMN1 and PTPRC is {len(cells_stmn1_ptprc)}\")\n",
    "print(f\" Number of cells expressing ANXA1 and PTPRC is {len(cells_anxa1_ptprc)}\")\n",
    "\n",
    "### print cells expressing single genes\n",
    "print(f\" Number of cells expressing STMN1 is {len(cells_stmn1)}\")\n",
    "print(f\" Number of cells expressing ANXA1 is {len(cells_anxa1)}\")\n",
    "print(f\" Number of cells expressing PTPRC is {len(cells_ptprc)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Number of cells expressing all 3 genes is 0\n",
      " Number of cells expressing STMN1 and ANXA1 is 1\n",
      " Number of cells expressing STMN1 and PTPRC is 0\n",
      " Number of cells expressing ANXA1 and PTPRC is 0\n",
      " Number of cells expressing STMN1 is 26\n",
      " Number of cells expressing ANXA1 is 9\n",
      " Number of cells expressing PTPRC is 0\n"
     ]
    }
   ],
   "source": [
    "### repeat with baysor cell centers\n",
    "cells_stmn1_b = baysor_seg_genes['cell'][baysor_seg_genes['gene']=='STMN1'].unique()\n",
    "cells_anxa1_b = baysor_seg_genes['cell'][baysor_seg_genes['gene']=='ANXA1'].unique()\n",
    "cells_ptprc_b = baysor_seg_genes['cell'][baysor_seg_genes['gene']=='PTPRC'].unique()\n",
    "\n",
    "### cells expressing all 3 genes\n",
    "cells_all_b = set(cells_stmn1_b) & set(cells_anxa1_b) & set(cells_ptprc_b)\n",
    "\n",
    "### cells expressing stmn1 & anxa1\n",
    "cells_stmn1_anxa1_b = set(cells_stmn1_b) & set(cells_anxa1_b)\n",
    "\n",
    "### cells expressing stmn1 & ptprc\n",
    "cells_stmn1_ptprc_b = set(cells_stmn1_b) & set(cells_ptprc_b)\n",
    "\n",
    "### cells expressing anxa1 & ptprc\n",
    "cells_anxa1_ptprc_b = set(cells_anxa1_b) & set(cells_ptprc_b)\n",
    "\n",
    "print(f\" Number of cells expressing all 3 genes is {len(cells_all_b)}\")\n",
    "print(f\" Number of cells expressing STMN1 and ANXA1 is {len(cells_stmn1_anxa1_b)}\")\n",
    "print(f\" Number of cells expressing STMN1 and PTPRC is {len(cells_stmn1_ptprc_b)}\")\n",
    "print(f\" Number of cells expressing ANXA1 and PTPRC is {len(cells_anxa1_ptprc_b)}\")\n",
    "\n",
    "### print cells expressing single genes\n",
    "print(f\" Number of cells expressing STMN1 is {len(cells_stmn1_b)}\")\n",
    "print(f\" Number of cells expressing ANXA1 is {len(cells_anxa1_b)}\")\n",
    "print(f\" Number of cells expressing PTPRC is {len(cells_ptprc_b)}\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### old and new segmentation files\n",
    "\n",
    "import json\n",
    "\n",
    "with open(out+'module6/script06_objects/segmentation/segmentation_polygons.json') as f:\n",
    "    baysor_boundaries = json.load(f)\n",
    "\n",
    "orig_boundaries = pd.read_csv(out+'module6/cell_boundaries_subset.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### plotting the segmentation results\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.patches import Polygon\n",
    "\n",
    "fig, ax = plt.subplots(1, 2, figsize=(10, 10))\n",
    "\n",
    "### plot the original segmentation\n",
    "\n",
    "grouped = orig_boundaries.groupby('cell_id')\n",
    "\n",
    "for cell_id, group in grouped:\n",
    "    group = pd.concat([group, group.iloc[[0:1]]])\n",
    "    plg = Polygon(group[['x', 'y']].values, edgecolor='black', facecolor='none')\n",
    "    ax[0].add_patch(plg)\n",
    "    \n",
    "ax[0].set_title('Original segmentation')\n",
    "\n",
    "\n",
    "### now plot tje baysor segmentation which is a json file\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "baysor_seg"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "xenium",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
